sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -339,8 +339,24 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
       case SqlBaseParser.SELECT =>
         // Regular select
 
+        // Broadcast hints
+        var withBroadcastedTable = relation
```

Comment:
For the first one, that could be possible. I did here because it's simple and I want to remove the unmatched MAPJOIN hints at the early stage. (You can see that the testcases) We will do the more complex real optimization in Optimizer layer.

For the second one, sure. That would better. I'll fix like that.